### PR TITLE
[AIRFLOW-3681] All GCP operators have now optional GCP Project ID

### PR DESCRIPTION
**Pull request by potiuk**
Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-3681

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:

The service account/connection to GCP can have an optional default
project id assigned. The project id can be specified explicitly
if we wish to override it, but when not specified (i.e. set to None)
the default one should be used instead.

This solution uses centralised decorator in order to have a common
place where retrieval logic is implemented for all operators.
Project_id is an important parameter of all GCP calls, therefore
in hooks it is left as always first positional parameter and it
has no default value. This has the added value that you have to
explicitly provide a value = None if you want to use default
project id.

On the other hand in operators, project_id is optional as in this
case you usually want to operate on default connection and project.

### Tests

- [x] My PR adds the following unit test cases:

* BigtableInstanceCreateTest.test_create_instance_that_exists_empty_project_id
* BigtableInstanceCreateTest.test_updating_cluster_but_instance_does_not_exists_empty_project_id
* BigtableInstanceCreateTest.test_updating_cluster_that_does_not_exists_empty_project_id
* BigtableInstanceDeleteTest.test_delete_execute_empty_project_id
* BigtableInstanceDeleteTest.test_deleting_instance_that_doesnt_exists_empty_project_id
* BigtableTableDeleteTest.test_deleting_table_that_doesnt_exists_empty_project_id
* BigtableTableCreateTest.test_creating_table_that_exists_empty_project_id
* GceInstanceStartTest.test_start_should_not_throw_ex_when_project_id_None
* GceInstanceStopTest.test_stop_should_not_throw_ex_when_project_id_none
* GceInstanceSetMachineTypeTest.test_set_machine_type_should_not_throw_ex_when_project_id_none
* GceInstanceGroupManagerUpdateTest.test_successful_instance_group_update_missing_project_id
* GceInstanceTemplateCopyTest.test_successful_copy_template_missing_project_id
* GcfFunctionDeployTest.test_empty_project_id_is_ok
* CloudSpannerTest.test_instance_create_missing_project_id
* CloudSpannerTest.test_instance_update_missing_project_id
* CloudSpannerTest.test_instance_query_missing_project_id
* CloudSpannerTest.test_database_update_missing_project_id
* CloudSpannerTest.test_database_delete_missing_project_id
* CloudSqlQueryValidationTest.test_create_operator_with_correct_parameters_project_id_missing
* CloudSqlTest.test_instance_create_missing_project_id
* CloudSqlTest.test_instance_patch_missing_project_id
* CloudSqlTest.test_instance_delete_missing_project_id
* CloudSqlTest.test_instance_db_create_missing_project_id
* CloudSqlTest.test_instance_db_patch_missing_project_id
* CloudSqlTest.test_instance_db_delete_missing_project_id
* CloudSqlTest.test_instance_export_missing_project_id
* CloudSqlTest.test_instance_import_missing_project_id

System tests updated with cases with project_id not set:
* example_gcp_sql_query.py
* example_gcp_spanner.sql
* example_gcp_sql.py
* example_gcp_spanner.py
* example_gcp_compute_igm.py
* example_gcp_compute.py
* example_gcp_bigtable_operators.py

### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
Documentation is updated. Example dags now include examples with and without project_id

### Code Quality

- [x] Passes `flake8`

### CODE REVIEW

**Comment by potiuk** [comment]
@jmcarp  -> this is the PR to change all GCP operators to have optional project_id. You might be interested.

File: airflow/contrib/hooks/gcp_api_base_hook.py [hunk 1]
```diff
@@ -144,11 +145,46 @@ def _get_field(self, f, default=None):
...
+            return func(self, *args, **kwargs)
+        return inner_wrapper
+
+    fallback_to_default_project_id = staticmethod(fallback_to_default_project_id)
+
+    def _get_project_id(self, project_id):
+        """
+        In case project_id is None, overrides it with default project_id from
+        the service account that is authorized.
+        :param project_id: project id to
```
Comment (kaxil) [inline airflow/contrib/hooks/gcp_api_base_hook.py:191]: Need an empty line above this

File: airflow/contrib/hooks/gcp_api_base_hook.py [hunk 2]
```diff
@@ -144,11 +145,46 @@ def _get_field(self, f, default=None):
...
     def project_id(self):
         return self._get_field('project')
+
+    def fallback_to_default_project_id(func):
+        """
+        Decorator that provides fallback for Google Cloud Platform project id. If
+        the project is None it will be replaced with the project_id from the
+        service account the Hook is authenticated with. Project id can be specified
+        either via project_id kwarg or via first parameter in positional args.
+        :param func: function to wrap
```
Comment (kaxil) [inline airflow/contrib/hooks/gcp_api_base_hook.py:164]: Need an empty line above this

File: airflow/contrib/hooks/gcp_api_base_hook.py [hunk 3]
```diff
@@ -144,11 +145,46 @@ def _get_field(self, f, default=None):
...
+        :return: result of the function call
+        """
+        @functools.wraps(func)
+        def inner_wrapper(self, *args, **kwargs):
+            if 'project_id' in kwargs:
+                kwargs['project_id'] = self._get_project_id(kwargs['project_id'])
+            else:
+                # assume project_id is the first non-keyword parameter if
```
Comment (kaxil) [inline airflow/contrib/hooks/gcp_api_base_hook.py]: The is a big assumption. we shouldn't assume this

File: airflow/contrib/hooks/gcp_bigtable_hook.py [hunk 4]
```diff
@@ -60,19 +64,23 @@ def get_instance(self, project_id, instance_id):
...
     def delete_instance(self, project_id, instance_id):
         """
         Deletes the specified Cloud Bigtable instance.
-        Raises google.api_core.exceptions.NotFound if the Cloud Bigtable instance does not exist.
+        Raises google.api_core.exceptions.NotFound if the Cloud Bigtable instance does
+        not exist.
 
-        :param project_id: The ID of the GCP project.
+        :param project_id: The ID of the GCP project. If set to None, default project_id
+            from service account is used.
```
Comment (kaxil) [inline airflow/contrib/hooks/gcp_bigtable_hook.py]: from the **Connection**

File: airflow/contrib/hooks/gcp_bigtable_hook.py [hunk 5]
```diff
@@ -90,7 +98,8 @@ def create_instance(self,
         Creates new instance.
 
         :type project_id: str
-        :param project_id: The ID of the GCP project.
+        :param project_id: The ID of the GCP project. If set to None, default project_id
+            from service account is used.
```
Comment (kaxil) [inline airflow/contrib/hooks/gcp_bigtable_hook.py]: from the Airflow connection, right?

File: airflow/contrib/hooks/gcp_function_hook.py [hunk 6]
```diff
@@ -46,11 +46,25 @@ def __init__(self,
         super(GcfHook, self).__init__(gcp_conn_id, delegate_to)
         self.api_version = api_version
 
+    @staticmethod
+    def _full_location(project_id, location):
+        """
+        Retrieve full location of the function in the form of
+        projects/<GCP_PROJECT_ID>/locations/<GCP_LOCATION>
+        :param project_id: The Google Cloud Project project_id where the function belongs.
```
Comment (kaxil) [inline airflow/contrib/hooks/gcp_function_hook.py:55]: Empty line needed above this

File: airflow/contrib/hooks/gcp_spanner_hook.py [hunk 7]
```diff
@@ -39,13 +37,14 @@ def __init__(self,
                  delegate_to=None):
         super(CloudSpannerHook, self).__init__(gcp_conn_id, delegate_to)
 
+    @GoogleCloudBaseHook.fallback_to_default_project_id
     def get_client(self, project_id):
-        # type: (str) -> Client
+        # noinspection LongLine
```
Comment (kaxil) [inline airflow/contrib/hooks/gcp_spanner_hook.py]: Can we remove this, please? It would also remove real long-line code

File: airflow/contrib/hooks/gcp_spanner_hook.py [hunk 8]
```diff
@@ -55,31 +54,36 @@ def get_client(self, project_id):
...
+        :param project_id: The ID of the  GCP project that owns the Cloud Spanner
+            database. if set to None, default project_id is used from service account.
         :type project_id: str
         :param instance_id: The ID of the Cloud Spanner instance.
         :type instance_id: str
         :return: Representation of a Cloud Spanner Instance. See:
             https://googleapis.github.io/google-cloud-python/latest/spanner/instance-api.html#google.cloud.spanner_v1.instance.Instance
         :rtype: object
         """
+        # noinspection PyUnresolvedReferences
```
Comment (kaxil) [inline airflow/contrib/hooks/gcp_spanner_hook.py]: Please remove all comments that disable style check. As we had a talk earlier that can be a separate PR where we resolve that issue

File: airflow/contrib/hooks/gcp_sql_hook.py [hunk 9]
```diff
@@ -682,15 +703,22 @@ class CloudSqlDatabaseHook(BaseHook):
...
 
-    def __init__(self, gcp_cloudsql_conn_id='google_cloud_sql_default'):
+    def __init__(self, gcp_cloudsql_conn_id='google_cloud_sql_default',
+                 default_gcp_project_id=None):
         super(CloudSqlDatabaseHook, self).__init__(source=None)
         self.gcp_cloudsql_conn_id = gcp_cloudsql_conn_id
         self.cloudsql_connection = self.get_connection(self.gcp_cloudsql_conn_id)
         self.extras = self.cloudsql_connection.extra_dejson
-        self.project_id = self.extras.get('project_id')
+        self.project_id = self.extras.get('project_id', default_gcp_project_id)
```
Comment (kaxil) [inline airflow/contrib/hooks/gcp_sql_hook.py:721]: It already gets the `project_id` from the connection. So, we don't need to add another default value

File: docs/howto/operator.rst [hunk 10]
```diff
@@ -730,6 +813,9 @@ Some arguments in the example DAG are taken from environment variables.
 Using the operator
 """"""""""""""""""
 
+You can create the operator with or without project id. If project id is missing
+it will be retrieved from the service account used. Both variants are shown:
```
Comment (kaxil) [inline docs/howto/operator.rst]: Connections???

**Review by kaxil — CHANGES_REQUESTED** [review]

File: airflow/contrib/hooks/gcp_api_base_hook.py (hunk 1 above)
Comment (potiuk) [reply airflow/contrib/hooks/gcp_api_base_hook.py:191]: OK

File: airflow/contrib/hooks/gcp_api_base_hook.py (hunk 2 above)
Comment (potiuk) [reply airflow/contrib/hooks/gcp_api_base_hook.py:164]: OK

File: airflow/contrib/hooks/gcp_api_base_hook.py (hunk 3 above)
Comment (potiuk) [reply airflow/contrib/hooks/gcp_api_base_hook.py]: Ok. I understand your concern. looked how it's done in apply_default for example and it's quite clever - it throws an exception if you pass positional args. I will try to follow the same pattern then.

File: airflow/contrib/hooks/gcp_bigtable_hook.py (hunk 4 above)
Comment (potiuk) [reply airflow/contrib/hooks/gcp_bigtable_hook.py]: Right. I change it everywhere.

File: airflow/contrib/hooks/gcp_bigtable_hook.py (hunk 5 above)
Comment (potiuk) [reply airflow/contrib/hooks/gcp_bigtable_hook.py]: Yeah. That was a mental shortcut.

File: airflow/contrib/hooks/gcp_function_hook.py (hunk 6 above)
Comment (potiuk) [reply airflow/contrib/hooks/gcp_function_hook.py:55]: I wish flake complained about it before :). Will do.

File: airflow/contrib/hooks/gcp_spanner_hook.py (hunk 7 above)
Comment (potiuk) [reply airflow/contrib/hooks/gcp_spanner_hook.py]: Removed all those, though I wish there was a more Intellij/PEP8/flake consistent way of pasting long URLs in docstrings :(

File: airflow/contrib/hooks/gcp_spanner_hook.py (hunk 8 above)
Comment (potiuk) [reply airflow/contrib/hooks/gcp_spanner_hook.py]: Done that. The PR for abstract class is in the making :)

File: airflow/contrib/hooks/gcp_sql_hook.py (hunk 9 above)
Comment (potiuk) [reply airflow/contrib/hooks/gcp_sql_hook.py:721]: Not really. This case is different. Cloud SQL query can have project_id specified in "extras" - so for example cloudsql://host:port/dbname?project_id=xx&sslcert=xxx....

If this is not specified, then we should take it from the GCP connection. There are two different connections passed to cloud SQL query operation - one is the "DB" cloudsql://" connection and the other is the GCP connection that is used to start Cloud SQL Proxy (if needed) - The 'cloud sql proxy' authenticates using a GCP connection (google_cloud_default by default).

The default project id is exactly for that reason - it is passed from the GCP connection as separate "default" parameter.

File: docs/howto/operator.rst (hunk 10 above)
Comment (potiuk) [reply docs/howto/operator.rst]: Yeah. That was a mental shortcut. In our environment we are using a service account to run GCP operation and we set project id from that service account as project_id parameter of the connection - so for use it's the service account's default project_id. But indeed in airflow terms it is a connection's project_id attribute value. Fixed

File: airflow/contrib/hooks/gcp_api_base_hook.py (hunk 3 above)
Comment (potiuk) [reply airflow/contrib/hooks/gcp_api_base_hook.py]: @kaxil - just wanted you to know that I am still working on it. I decided to do it "properly" - i.e. change all project_ids to be keyword parameters and have them in hook methods as last parameters with default = None. Thanks to our automated system tests i will be able to test it for all operators e-2-e not only unit tests (I am doing it as I am replacing service-by-service). 

I also realised that since I am touching pretty much every hook method, it will be safer to add unit tests for all the hooks. This way I can test if project_id is indeed optional and whether it can be overridden for all hook methods. 
Lack of hook unit tests  was a big missing point that we had in the implementation, and It seems that i can do it rather quickly now - it's pretty tedious but low-risk, high return.

I am done with almost all services - working on Spanner and ACL now (I have working tests for all other services). I hope tomorrow I will send revised PR - with literally hundreds more unit tests for hooks and truly optional project_id. I hope it can make it to the release then :)

**Comment by potiuk** [comment]
All is done now @kaxil : This is a monster change, but it's mostly in tests. The change includes changing the project_id to be keyword based and Added 150+(!) new unit test cases for all hooks in GCP.

I also corrected all places where the hooks were returning True if successful and raising Exception if there was a problem. It was mentioned before and we agreed it made little sense - now the hooks return Nothing.

I also passed it through automated system tests with real GCP project and the tests are all green :). 

Here is the result of our automated build for System tests and relevant unit test cases.

https://storage.googleapis.com/polidea-airflow-builds/4846b9c3-8b0d-418f-b302-52980928b284/index.html

**Comment by potiuk** [comment]
@jmcarp  -> you might take a look as well. This one is vastly improved version of optional project_id - now in hooks all the parameters are forced to be keyword parameters and project_id is still updated with decorator, and I have added many, many tests to cover hooks not only operators/system tests.

**Review by kaxil — APPROVED** [review]

**Comment by kaxil** [comment]
Thanks @potiuk

**Comment by potiuk** [comment]
Thanks @kaxil !
